Use PHP functions to modify query strings

leofeyer opened the pull request:
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #3392
| Docs PR or issue | -

Tested with the page structure of the official demo.

### Before

<img width="435" alt="" src="https://user-images.githubusercontent.com/1192057/131829916-5e3ec15a-9d85-49f6-98d4-3a2ea6129b6e.png">

### After

<img width="433" alt="" src="https://user-images.githubusercontent.com/1192057/131829942-3a2d53b4-1143-4f7e-a18d-663a302e6edc.png">

Needs to be backported to Contao 4.9 if we agree that changing the major version is OK.

leofeyer committed:
Update league/uri-components to version 2

leofeyer committed:
Only create the query object once

Toflar commented on core-bundle/src/Resources/contao/library/Contao/Controller.php:
```diff
@@ -1070,29 +1075,26 @@ public static function generateMargin($arrValues, $strType='margin')
-		$query = new Query(Environment::get('queryString'));
+		if (static::$objQuery === null)
+		{
+			static::$objQuery = Query::createFromRFC3986(Environment::get('queryString'));
```
Why is this not from `Request` like the `_contao_referer_id` as well?

leofeyer replied:
Because the `_contao_referer_id` part is new and the rest is legacy code? I am fine with changing this, though.

Maybe we can get rid of `leagure/uri-components` completely if we use a combination of Symfony's `HeaderUtils::parseQuery()`, some custom logic and `http_build_query()` (like Symfony does in `Request::normalizeQueryString()`)?

leofeyer replied:
See 6d1f50dd22e02bc20f038b4a56b639b7352f146d

fritzmg commented:
I am wondering in general why we use `league/uri-components` for `Backend::addToUrl`, but not for `Frontend::addToUrl` as well (or vice versa).

leofeyer commented:
> The problem isn't the component but the ridiculous amount of calls to it. Likely unwanted recursion. 
> _Originally posted by @m-vo in https://github.com/contao/contao/issues/3392#issuecomment-910156791_

I don't think so. 35 pages with 8 buttons each will trigger 280 calls alone. Together with module links and global buttons, 302 calls are very likely.

leofeyer commented:
> I am wondering in general why we use `league/uri-components` for `Backend::addToUrl`, but not for `Frontend::addToUrl` as well (or vice versa).

Unrelated discussion. Please open a separate issue.

fritzmg commented:
> Unrelated discussion. Please open a separate issue.

It's not unrelated if the custom implementation in `Frontend::addToUrl` without `league/uri-components` is faster than the one in `Backend::addToUrl`, that's why I brought it up.

leofeyer committed:
Add unit tests

leofeyer committed:
Get the query string from the request object

Toflar commented on core-bundle/tests/Contao/ControllerTest.php:
```diff
@@ -348,4 +349,94 @@ public function testAddImageToTemplateLogsAndCreatesFallbackDataWhenNoFigureIsBu
+     */
+    public function testAddToUrlWithoutQueryString(): void
+    {
+        define('TL_SCRIPT', '');
```
If we're already working on this method we could also get rid of this deprecated constant, no?

leofeyer replied:
That is also not related to this PR. Maybe create a separate PR and fix it everywhere instead of only in one place?

aschempp replied:
I have added it to the Contao 5 project ideas.

leofeyer replied:
We don't have to wait for Contao 5. There is already an alternative to using `TL_SCRIPT` IIRC, so we could change this right away. It was just out of the scope of this PR.

leofeyer committed:
Try it without league/uri

leofeyer commented:
I have added a variant without `league/uri-components` in 16395a221e69fe55cf6f5f68eb5059c1a42a0a9e. The performance is only slightly better than with `league/uri-components` in version 2, , therefore we should ask us if we rather want to use a well-tested third-party library or a custom function (I have added unit tests for it, too). @contao/developers WDYT?

leofeyer committed:
Fix the CI chain

leofeyer committed:
Save the StringUtil::ampersand() call

ausi commented:
> if we rather want to use a well-tested third-party library or a custom function

In this case I would use the custom function as the problem it solves is not complex enough to justify the use of a library IMO.

leofeyer committed:
Replace the other occurrences of league/uri as well

leofeyer commented on core-bundle/src/Resources/contao/pages/PageLogout.php:
```diff
@@ -56,14 +55,21 @@ public function getResponse($objPage)
-		$query = $query->merge('redirect=' . $strRedirect);
+		$pairs['redirect'] = $strRedirect;
+
+		$request->server->set('QUERY_STRING', http_build_query($pairs, '', '&', PHP_QUERY_RFC3986));
```
Not sure whether modifying the request object is the best solution here?

ausi replied:
Maybe we could use `parse_url(…, PHP_URL_QUERY)` instead of the Request class and string replace the query afterwards?

leofeyer replied:
The "replacing the query afterwards" is the problematic part here. The Symfony `Request` class is able to create a URI from the `parse_url()` parts, that's why I have used it here. Do you know a better way?

ausi replied:
Something like this? (untested)
```php
$url = preg_replace('(\?'.preg_quote($oldQuery).'(?=#|$))', '?'.$newQuery, $url);
```

leofeyer replied:
I don't think that this is "better" than using the request object. 🙈 @contao/developers WDYT?

Toflar replied:
`Request::create()` accepts `$_SERVER` like arguments, so you could pass it in there.

leofeyer replied:
I have already tried this with `$request->duplicate(…, ['QUERY_STRING' => '…'])`, however, this will replace the `$server` variable instead of merging my array.

ausi replied:
> I don't think that this is "better" than using the request object.

You’re right, it is probably not better ☺️

The current solution is fine then I think. 👍

leofeyer replied:
I have made a small adjustment in 8da7cde8cadf4a44a2959adb164060027bfb605c, so we no longer have to overwrite the `QUERY_STRING` variable.

ausi commented on core-bundle/src/Resources/contao/library/Contao/Controller.php:
```diff
@@ -1070,29 +1074,53 @@ public static function generateMargin($arrValues, $strType='margin')
+		}
 
 		// Add the referer ID
 		if (isset($_GET['ref']) || ($strRequest && $blnAddRef))
```
Should we ask the request here instead of `$_GET`?

leofeyer replied:
Yes. Changed in f013fab6a704077a2e9c6890fa979fc6ed084c7c.

ausi commented on core-bundle/src/Resources/contao/library/Contao/System.php:
```diff
@@ -400,10 +399,14 @@ public static function getReferer($blnEncodeAmpersands=false, $strTable=null)
+				{
+					unset($pairs[$param]);
+				}
+
```
Should we return without the `?` if the whole query string was unset?
```suggestion

				if (!$pairs) 
				{
					return $path;
				}

```

leofeyer replied:
Merged in cbae1469f9fa05aa3672f68e518ad86cdd77d435.

leofeyer committed:
Do not create a query string if there are no pairs

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

leofeyer committed:
Fix the CI chain

leofeyer committed:
Use the request object instead of $_GET

Toflar reviewed (dismissed):
Thanks for the effort ❤️

leofeyer committed:
Do not overwrite the QUERY_STRING variable
